### PR TITLE
Gitignore for GO lang

### PR DESCRIPTION
**Pull request by K-Kumar-01**
## Description
Created a new practice that checks correctly set .gitignore for Go lang.

## Motivation and Context

The issue related to the PR is linked [here](https://github.com/DXHeroes/dx-scanner/issues/407)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated the documentation accordingly.
- [ ] I have added new practice to practice list in README.md.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    expect(evaluated).toEqual(PracticeEvaluationResult.unknown);
+  });
+
+  describe('Fixer', () => {
+    afterEach(async () => {
+      containerCtx.virtualFileSystemService.clearFileSystem();
+    });
+
+    it('Does not change correct .gitignore', async () => {
+      const gitignore = `${basicGitignore}\npackage-lock.json\n`;
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts:70]: There is no `package-lock.json` in Go lang. Please, change it.

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import { ReportDetailType } from '../../reporters/ReporterData';
+import { FixerContext } from '../../contexts/fixer/FixerContext';
+
+@DxPractice({
+  id: 'Go.GitignoreCorrectlySet',
+  name: 'Set .gitignore Correctly',
+  impact: PracticeImpact.high,
+  suggestion: 'Set patterns in the .gitignore as usual.',
+  reportOnlyOnce: true,
+  url: 'https://github.com/github/gitignore/blob/master/Node.gitignore',
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:18]: Change the url for Go .`gitignore`

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      return gitignoreFile
+        .toString()
+        .split(/\r?\n/)
+        .filter((content) => content.trim() !== '' && !content.startsWith('#'));
+    };
+    const content = await ctx.root.fileInspector.readFile('.gitignore');
+    const parsedGitignore = parseGitignore(content);
+    this.parsedGitignore = parsedGitignore;
+
+    // node_modules
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:43]: This is not nod_modules. Change the comment

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    const binaryFilesRegex = parsedGitignore.find((value: string) =>
+      /.exe$|.exe~$|.dll$|.so$|.dylib$|.test$/.test(value),
+    );
+    // misc
+    const coverageRegex = parsedGitignore.find((value: string) =>
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:48]: I'm not Go developer, but are `.out` files used for coverage?

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    this.setData();
+    return PracticeEvaluationResult.notPracticing;
+  }
+
+  async fix(ctx: FixerContext) {
+    const inspector = ctx.fileInspector?.basePath
+      ? ctx.fileInspector
+      : ctx.root.fileInspector;
+    if (!inspector) return;
+    // node_modules
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:68]: same as above

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    const binaryFilesRegex = this.parsedGitignore.find((value: string) =>
+      /.exe$|.exe~$|.dll$|.so$|.dylib$|.test$/.test(value),
+    );
+    // misc
+    const coverageRegex = this.parsedGitignore.find((value: string) =>
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:73]: same as above

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        .split(/\r?\n/)
+        .filter((content) => content.trim() !== '' && !content.startsWith('#'));
+    };
+    const content = await ctx.root.fileInspector.readFile('.gitignore');
+    const parsedGitignore = parseGitignore(content);
+    this.parsedGitignore = parsedGitignore;
+
+    // node_modules
+    const binaryFilesRegex = parsedGitignore.find((value: string) =>
+      /.exe$|.exe~$|.dll$|.so$|.dylib$|.test$/.test(value),
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:45]: See the comment here https://github.com/DXHeroes/dx-scanner/issues/407#issuecomment-714479986

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  async fix(ctx: FixerContext) {
+    const inspector = ctx.fileInspector?.basePath
+      ? ctx.fileInspector
+      : ctx.root.fileInspector;
+    if (!inspector) return;
+    // node_modules
+    const binaryFilesRegex = this.parsedGitignore.find((value: string) =>
+      /.exe$|.exe~$|.dll$|.so$|.dylib$|.test$/.test(value),
```
Comment (adelkahomolova) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.ts:70]: See the comment here https://github.com/DXHeroes/dx-scanner/issues/407#issuecomment-714479986

**Review by adelkahomolova — CHANGES_REQUESTED** [review]

File: src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts (hunk 1 above)
Comment (K-Kumar-01) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts:70]: we need to simply remove it or add any other file likt that of `package-lock.json`.

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts (hunk 4 above)
Comment (K-Kumar-01) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.ts:64]: [link](https://github.com/github/gitignore/blob/master/Go.gitignore) tells it as `Output of the go coverage tool, specifically when used with LiteIDE`.
so i used the word coverage

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts (hunk 8 above)
Comment (K-Kumar-01) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.ts]: The new PR doesn't have this issue. I changed this area

File: src/practices/Go/GoGitignoreCorrectlySetPractice.ts (hunk 7 above)
Comment (K-Kumar-01) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.ts]: Changed this in the new PR made

File: src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts (hunk 1 above)
Comment (adelkahomolova) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts]: You can just remove the `package-lock.json`

**Review by adelkahomolova — APPROVED** [review]

**Comment by adelkahomolova** [comment]
@K-Kumar-01 , I appreciate the time you invest in our repository. I hope you enjoyed the contributing to DX-Scanner! It's approved by me. I'll just wait for a review from my colleague as we think it's the best practice to have 2 approvals to keep the high quality of code.
If you'll want to continue in contributing in open-source, check our other issues on [dx-scanner](https://github.com/DXHeroes/dx-scanner/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc). If you're good at writing, (or you just want to self-education) check our open-source [Knowledge Base](https://developerexperience.io/) (here are the [issues](https://github.com/DXHeroes/knowledge-base-content/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc) )
Take care of yourself in these complicated days! ❤️

File: src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  it('Returns practicing if the .gitignore is set correctly', async () => {
+    containerCtx.virtualFileSystemService.setFileSystem({
+      '.gitignore': gitignoreContent,
+    });
+
+    const evaluated = await practice.evaluate(containerCtx.practiceContext);
+    expect(evaluated).toEqual(PracticeEvaluationResult.practicing);
+  });
+
+  it('Returns notPracticing if there the .gitignore is NOT set correctly', async () => {
```
Comment (Jakub-Vacek) [inline src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts:46]: Returns notPracticing if the .gitignore is NOT set correctly

Comment (K-Kumar-01) [reply src/practices/Go/GoGitignoreCorrectlySetPractice.spec.ts]: Made the requested change in the latest push

**Review by Jakub-Vacek — APPROVED** [review]

**Comment by adelkahomolova** [comment]
Screenshot 2020-10-26 at 14 20 56
Hi, @K-Kumar-01 , I noticed one test, which is failing. Can you fix it, please? :)

**Comment by K-Kumar-01** [comment]
> Screenshot 2020-10-26 at 14 20 56
> 
> Hi, @K-Kumar-01 , I noticed one test, which is failing. Can you fix it, please? :)

I have changed the sentence regrading this test case.
Let me know if any other change is needed.

**Comment by adelkahomolova** [comment]
@K-Kumar-01 the test is not failing because there is bad text in the name of the test. You need to fix the code. The test is failing. I can't merge it since there is something bad.

**Comment by K-Kumar-01** [comment]
> @K-Kumar-01 the test is not failing because there is bad text in the name of the test. You need to fix the code. The test is failing. I can't merge it since there is something bad.

OK I will see to it and try to fix the problem. 
I will let you know once i have made the changes

**Comment by K-Kumar-01** [comment]
@adelkahomolova 
I am not able to understand where i have to change the code as it is similar to other practices.
Can you tell me what should i change.
I did this though `const basicGitignore = `*.exe\n*.exe~\n*.dll\n*.so\n*.dylib\n*.test\n*.out\nvendor/\n`;`
thinking that newline characters may be a cause for the failing test cases.

**Comment by adelkahomolova** [comment]
Hi, @K-Kumar-01 , I looked at it. The problem is that you check also the `vendor`, which is commented in the `.gitignore`. All comments are filtered out in the `parseGitignore()` method. You shouldn't check the vendor string.

**Comment by K-Kumar-01** [comment]
@adelkahomolova The comment above vendor stated dependency directories. so i thought it shouldn't be a comment.
I have made the changes in new PR. 
Let me know if there is any change needed.

**Comment by adelkahomolova** [comment]
Thank you for the changes, @K-Kumar-01 , I've already merged your PR :)

**Pull request merged**